File: rust/src/core/session.rs
//! Session-based auto-grouping for dexcost.
//!
//! Groups related LLM and HTTP calls into a single task without requiring
//! explicit task wrappers. Uses a thread-safe `SessionManager` that tracks
//! sessions by thread/context ID with idle timeout finalization.

use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};

use chrono::Utc;

use crate::core::context::{get_current_task, get_dexcost_context_sync};
use crate::core::models::{Task, TaskStatus};

/// Entry in the session map — holds a task and the last activity timestamp.
struct SessionEntry {
    task: Task,
    last_activity: Instant,
}

/// Manages auto-created session tasks for grouping cost events.
///
/// When no explicit task is active, the session manager creates a session task
/// and returns it so that subsequent LLM and HTTP calls are grouped together.
pub struct SessionManager {
    sessions: Mutex<HashMap<u64, SessionEntry>>,
    idle_timeout: Duration,
}

impl SessionManager {
    /// Creates a new `SessionManager` with the given idle timeout.
    ///
    /// Sessions that have had no activity for longer than `idle_timeout` will
    /// be finalized by [`finalize_idle_sessions`](Self::finalize_idle_sessions).
    pub fn new(idle_timeout: Duration) -> Self {
        Self {
            sessions: Mutex::new(HashMap::new()),
            idle_timeout,
        }
    }

    /// Returns the active task from context, or creates a new session task.
    ///
    /// If an explicit task is already active in the current context (via
    /// task-local storage), that task is returned unchanged. Otherwise, a
    /// session task is created (or reused) for the current thread.
    ///
    /// # Arguments
    ///
    /// * `call_type` - Description of the call (e.g. `"llm_call"`, `"http_call"`).
    ///   Used as the `task_type` when creating a new session task if no ambient
    ///   context provides an agent name.
    pub fn get_or_create_session(&self, call_type: &str) -> Task {
        // If an explicit task is already active, use it and refresh activity.
        if let Some(existing) = get_current_task() {
            let ctx_id = Self::context_id();
            let mut sessions = self.sessions.lock().unwrap_or_else(|e| {
                eprintln!("[dexcost] mutex poisoned, recovering: {}", e);
                e.into_inner()
            });
            if let Some(entry) = sessions.get_mut(&ctx_id) {
                entry.last_activity = Instant::now();
            }
            return existing;
        }

        let ctx_id = Self::context_id();
        let mut sessions = self.sessions.lock().unwrap_or_else(|e| {
            eprintln!("[dexcost] mutex poisoned, recovering: {}", e);
            e.into_inner()
        });

        // Reuse existing session for this context.
        if let Some(entry) = sessions.get_mut(&ctx_id) {
            entry.last_activity = Instant::now();
            return entry.task.clone();
        }

        // Create a new session task.
        let task_type = if call_type.is_empty() {
            "agent_session".to_string()
        } else {
            call_type.to_string()
        };

        let mut task = Task::new(&task_type);
        task.metadata
            .insert("session".to_string(), serde_json::Value::Bool(true));
        task.metadata.insert(
            "initiated_by".to_string(),
            serde_json::Value::String(call_type.to_string()),
        );

        // Try to pull customer_id / project_id from ambient context.
        // We avoid tokio::task::block_in_place here because it can deadlock
        // if the tokio runtime is single-threaded. Instead, use try_read()
        // on the underlying RwLock (non-blocking, best-effort).
        // The ambient context is a nice-to-have, not critical.
        if let Some(ctx) = get_dexcost_context_sync() {
            task.customer_id = ctx.customer_id;
            task.project_id = ctx.project_id;
        }

        sessions.insert(
            ctx_id,
            SessionEntry {
                task: task.clone(),
                last_activity: Instant::now(),
            },
        );

        task
    }

    /// Closes sessions that have been idle for longer than the configured timeout.
    ///
    /// Returns the list of finalized session tasks with status set to `Success`
    /// and `ended_at` set to the current time.
    pub fn finalize_idle_sessions(&self) -> Vec<Task> {
        let now = Instant::now();
        let mut finalized = Vec::new();

        let mut sessions = self.sessions.lock().unwrap_or_else(|e| {
            eprintln!("[dexcost] mutex poisoned, recovering: {}", e);
            e.into_inner()
        });
        let idle_ids: Vec<u64> = sessions
            .iter()
            .filter(|(_, entry)| now.duration_since(entry.last_activity) >= self.idle_timeout)
            .map(|(id, _)| *id)
            .collect();

        for ctx_id in idle_ids {
            if let Some(mut entry) = sessions.remove(&ctx_id) {
                entry.task.status = TaskStatus::Success;
                entry.task.ended_at = Some(Utc::now());
                finalized.push(entry.task);
            }
        }

        finalized
    }

    /// Returns the number of currently active sessions.
    pub fn active_session_count(&self) -> usize {
        let sessions = self.sessions.lock().unwrap_or_else(|e| {
            eprintln!("[dexcost] mutex poisoned, recovering: {}", e);
            e.into_inner()
        });
        sessions.len()
    }

    /// Removes all tracked sessions (primarily for testing).
    pub fn clear(&self) {
        let mut sessions = self.sessions.lock().unwrap_or_else(|e| {
            eprintln!("[dexcost] mutex poisoned, recovering: {}", e);
            e.into_inner()
        });
        sessions.clear();
    }

    /// Returns a unique context identifier for the current thread.
    fn context_id() -> u64 {
        // Use the raw thread ID as the context key.
        let id = std::thread::current().id();
        // Thread::id() -> ThreadId; use its Debug representation to extract a
        // stable u64. ThreadId(N) is guaranteed unique for the thread's lifetime.
        let debug = format!("{:?}", id);
        // Extract the numeric portion.
        debug
            .chars()
            .filter(|c| c.is_ascii_digit())
            .collect::<String>()
            .parse::<u64>()
            .unwrap_or(0)
    }
}

// ---------------------------------------------------------------------------
// Module-level singleton
// ---------------------------------------------------------------------------

static SESSION_MANAGER: std::sync::OnceLock<SessionManager> = std::sync::OnceLock::new();

/// Returns the global session manager, creating it with a 30-second idle
/// timeout if it has not been initialized yet.
pub fn get_session_manager() -> &'static SessionManager {
    SESSION_MANAGER.get_or_init(|| SessionManager::new(Duration::from_secs(30)))
}

/// Resets the global session manager (for testing). Clears all sessions but
/// does not remove the singleton — it will be reused on the next call.
pub fn reset_session_manager() {
    if let Some(mgr) = SESSION_MANAGER.get() {
        mgr.clear();
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::Duration;

    #[test]
    fn test_new_session_created() {
        let mgr = SessionManager::new(Duration::from_secs(30));
        let task = mgr.get_or_create_session("llm_call");
        assert_eq!(task.task_type, "llm_call");
        assert_eq!(task.status, TaskStatus::Pending);
        assert!(!task.task_id.is_empty());
        assert_eq!(mgr.active_session_count(), 1);
    }

    #[test]
    fn test_reuses_existing_session() {
        let mgr = SessionManager::new(Duration::from_secs(30));
        let first = mgr.get_or_create_session("llm_call");
        let second = mgr.get_or_create_session("http_call");
        // Same thread -> same session reused.
        assert_eq!(first.task_id, second.task_id);
        assert_eq!(mgr.active_session_count(), 1);
    }

    #[test]
    fn test_session_metadata() {
        let mgr = SessionManager::new(Duration::from_secs(30));
        let task = mgr.get_or_create_session("llm_call");
        assert_eq!(
            task.metadata.get("session"),
            Some(&serde_json::Value::Bool(true))
        );
        assert_eq!(
            task.metadata.get("initiated_by"),
            Some(&serde_json::Value::String("llm_call".to_string()))
        );
    }

    #[test]
    fn test_finalize_idle_sessions() {
        let mgr = SessionManager::new(Duration::from_millis(1));
        let _task = mgr.get_or_create_session("llm_call");
        assert_eq!(mgr.active_session_count(), 1);

        // Wait for the session to become idle.
        std::thread::sleep(Duration::from_millis(10));

        let finalized = mgr.finalize_idle_sessions();
        assert_eq!(finalized.len(), 1);
        assert_eq!(finalized[0].status, TaskStatus::Success);
        assert!(finalized[0].ended_at.is_some());
        assert_eq!(mgr.active_session_count(), 0);
    }

    #[test]
    fn test_finalize_does_not_remove_active() {
        let mgr = SessionManager::new(Duration::from_secs(300));
        let _task = mgr.get_or_create_session("llm_call");

        let finalized = mgr.finalize_idle_sessions();
        assert!(finalized.is_empty());
        assert_eq!(mgr.active_session_count(), 1);
    }

    #[test]
    fn test_clear() {
        let mgr = SessionManager::new(Duration::from_secs(30));
        let _task = mgr.get_or_create_session("llm_call");
        assert_eq!(mgr.active_session_count(), 1);
        mgr.clear();
        assert_eq!(mgr.active_session_count(), 0);
    }

    #[test]
    fn test_empty_call_type_defaults_to_agent_session() {
        let mgr = SessionManager::new(Duration::from_secs(30));
        let task = mgr.get_or_create_session("");
        assert_eq!(task.task_type, "agent_session");
    }

    #[test]
    fn test_global_session_manager() {
        reset_session_manager();
        let mgr = get_session_manager();
        assert_eq!(mgr.active_session_count(), 0);
    }
}

File: typescript/src/core/session.ts
/**
 * Session-based auto-grouping.
 *
 * Groups LLM + HTTP calls into one task per execution context
 * without requiring explicit tracker.track() wrappers.
 */

import { randomUUID } from "node:crypto";
import type { Task } from "./models.js";
import { createTask } from "./models.js";
import { getCurrentTask, runWithTask, getContext } from "./context.js";
import {
  NetworkAccountant,
  registerAccountant,
} from "../adapters/network-accountant.js";
import { finalizeTaskNetwork } from "./network-finalize.js";
import type { EventBuffer } from "../transport/buffer.js";

// ---------------------------------------------------------------------------
// Session tracking
// ---------------------------------------------------------------------------

interface SessionInfo {
  task: Task;
  lastActivityAt: number;
}

/** Idle threshold in milliseconds (30 seconds). */
const IDLE_THRESHOLD_MS = 30_000;

/**
 * Grouping key for calls made with no ambient DexcostContext set.
 *
 * The Python SDK groups sessions by `threading.get_ident()` — one session
 * per thread, reused across calls until 30s idle. Node is single-threaded,
 * so the equivalent is one process-wide session. When the app DOES set an
 * ambient context (`setContext(...)` per request/job), the context object's
 * identity is the grouping key instead, giving per-request sessions.
 */
const GLOBAL_SESSION_KEY: unknown = Symbol("dexcost.global_session");

/**
 * SessionManager — creates and reuses session tasks automatically.
 *
 * When no explicit task is set via `tracker.track()`, the session manager
 * creates a session task keyed on the ambient DexcostContext (or a global
 * key when none is set) and REUSES it for subsequent calls until the
 * session goes idle. Pre-fix, sessions were keyed by their own taskId and
 * never re-bound, so every un-wrapped HTTP/LLM call produced a brand-new
 * single-call "agent_session" task — no grouping ever happened (the
 * attribution scatter seen as many isolated 1-call tasks).
 */
export class SessionManager {
  /** Grouping key (context object | GLOBAL_SESSION_KEY) → session. */
  private _sessions: Map<unknown, SessionInfo> = new Map();
  /** taskId → session, for activity updates and test lookups. */
  private _byTaskId: Map<string, SessionInfo> = new Map();

  /** Resolve the grouping key for the current async context. */
  private _sessionKey(): unknown {
    return getContext() ?? GLOBAL_SESSION_KEY;
  }

  /** Create a session task for the current ambient context and index it. */
  private _createSession(callType: string, buffer: EventBuffer, key: unknown): SessionInfo {
    const ctx = getContext();
    const taskType = ctx?.agent ?? "agent_session";

    const task = createTask({
      taskId: randomUUID(),
      taskType,
      customerId: ctx?.customerId,
      projectId: ctx?.projectId,
      metadata: { session: true, initiatedBy: callType },
    });

    // Register a NetworkAccountant so the patched fetch attributes the
    // session's bytes to this task; drained + egress-priced when the
    // session is finalized (idle sweep or shutdown).
    registerAccountant(task.taskId, new NetworkAccountant());

    buffer.upsertTask(task);

    const info: SessionInfo = { task, lastActivityAt: Date.now() };
    this._sessions.set(key, info);
    this._byTaskId.set(task.taskId, info);
    return info;
  }

  /**
   * Get the currently-active task, or create a new session task if none exists.
   *
   * If a task is already active in the AsyncLocalStorage context (from an
   * explicit `tracker.track()` call), it is returned as-is. Otherwise the
   * session for the current grouping key is reused (creating it on first
   * call).
   *
   * @param callType  Label for the type of call (e.g. "http", "llm")
   * @param buffer    EventBuffer to persist the new task
   * @returns The active or newly-created Task
   */
  getOrCreateSession(callType: string, buffer: EventBuffer): Task {
    const existing = getCurrentTask();
    if (existing !== undefined) {
      // Update last activity for existing session if we are tracking it
      const sessionInfo = this._byTaskId.get(existing.taskId);
      if (sessionInfo) {
        sessionInfo.lastActivityAt = Date.now();
      }
      return existing;
    }

    const key = this._sessionKey();
    const info = this._sessions.get(key);
    if (info) {
      info.lastActivityAt = Date.now();
      return info.task;
    }

    // NOTE: intentionally NOT calling setCurrentTask(task) here.
    // enterWith() leaks the task into the remainder of the async chain,
    // causing subsequent unwrapped calls to inherit a stale session task.
    // Callers should use runInSession() (which scopes via runWithTask);
    // cross-call reuse instead comes from the grouping-key map above.
    return this._createSession(callType, buffer, key).task;
  }

  /**
   * Run a function within a session context.
   *
   * Creates (or reuses) a session task if none is active, then runs `fn`
   * within that task's AsyncLocalStorage context.
   */
  runInSession<T>(callType: string, buffer: EventBuffer, fn: () => T): T {
    const existing = getCurrentTask();
    if (existing !== undefined) {
      const sessionInfo = this._byTaskId.get(existing.taskId);
      if (sessionInfo) {
        sessionInfo.lastActivityAt = Date.now();
      }
      return fn();
    }

    const key = this._sessionKey();
    let info = this._sessions.get(key);
    if (info) {
      info.lastActivityAt = Date.now();
    } else {
      info = this._createSession(callType, buffer, key);
    }

    return runWithTask(info.task, fn);
  }

  /**
   * Finalize one session: drain its NetworkAccountant into byte aggregates
   * + egress dollars, mark it terminal, and persist.
   */
  private _finalizeSession(info: SessionInfo, buffer?: EventBuffer): void {
    if (info.task.status === "pending") {
      info.task.status = "success";
      info.task.endedAt = new Date();
    }
    try {
      // Drain regardless of prior status so the accountant registry entry
      // never leaks; stamps networkCostUsd for the pending→success case.
      finalizeTaskNetwork(info.task, buffer);
    } catch {
      // Fail-silent: pricing problems must never break the sweep.
    }
    if (buffer) {
      buffer.upsertTask(info.task);
    }
  }

  /**
   * Finalize sessions that have been idle for longer than 30 seconds.
   *
   * Sets their status to "success", drains network bytes into egress cost,
   * and stamps endedAt. Removes them from the active session map. When a
   * `buffer` is provided, re-persists each finalized task so the updated
   * status reaches the push cycle.
   */
  finalizeIdleSessions(buffer?: EventBuffer): void {
    const now = Date.now();
    const toRemove: unknown[] = [];

    for (const [key, session] of this._sessions) {
      if (now - session.lastActivityAt > IDLE_THRESHOLD_MS) {
        this._finalizeSession(session, buffer);
        toRemove.push(key);
      }
    }

    for (const key of toRemove) {
      const info = this._sessions.get(key);
      this._sessions.delete(key);
      if (info) this._byTaskId.delete(info.task.taskId);
    }
  }

  /**
   * Finalize ALL active sessions regardless of idle time.
   *
   * Called during shutdown (`close` / `closeAsync`) to ensure no session
   * tasks are left in "pending" status.
   */
  finalizeAllSessions(buffer?: EventBuffer): void {
    for (const [, session] of this._sessions) {
      this._finalizeSession(session, buffer);
    }
    this._sessions.clear();
    this._byTaskId.clear();
  }

  /** Number of active sessions. */
  get activeSessionCount(): number {
    return this._sessions.size;
  }

  /** Get a session's task by ID (for testing). */
  getSession(taskId: string): Task | undefined {
    return this._byTaskId.get(taskId)?.task;
  }
}
